WIP: Active-Active streams edits

banker opened the pull request.

Paz-Yanover commented on content/rs/developing/crdbs/developing-streams-crdb.md:
```diff
@@ -1,278 +1,265 @@
+| _t5_   | `XINFO GROUPS x` <br/>**→ [g1, g2]** | `XINFO GROUPS x` <br/>**→ [g1, g2]** |
 
-Here is an example of a concurrent XADD case:
+Here is an example of creating two consumer groups concurrently:
```
Creating consumer groups concurrently is a nice example, but I think that XADD-ing messages in parallel is much more interesting and useful

banker replied:
There are examples of XADDing messages in parallel elsewhere in this doc. This example is in the context of consumer groups specifically.

banker commented:
@Paz-Yanover @AlonMagrafta @bmansheim If there are no major objections, I'd like to merge this ASAP.

banker commented on content/rs/developing/crdbs/developing-streams-crdb.md:
```diff
@@ -1,278 +1,265 @@
-| t3   |                                 | XADD x 100-2 f1 v1              |
-| t4   | --- sync ---                    | --- sync ---                    |
-| t5   | XRANGE x - + --> [100-1, 100-2] | XRANGE x - + --> [100-1, 100-2] |
+| _t1_   | `XADD x 100-1 f1 v1`              |                                 |
```
I agree. Will fix.

banker commented on content/rs/developing/crdbs/developing-streams-crdb.md:
```diff
@@ -1,278 +1,265 @@
+### Preserving ID uniqueness
 
-Because of the asynchronous nature of Active-Active databases, the following scenario is possible:
+Because Active-Active databases replicate asynchronously, it's possible to create streams with duplicate IDs:
```
I agree. Will rework this.

banker commented on content/rs/developing/crdbs/developing-streams-crdb.md:
```diff
@@ -1,278 +1,265 @@
+The default and recommended mode is _strict_, which prevents duplicate IDs. It's generally important to avoid duplicate IDs because:
+
+1. XDEL, XCLAIM, and other commands may affect more than one entry when duplicate IDs are present
+1. Entries with duplicate IDs may be removed if a database is exported or renamed
```
Will fix.

banker commented on content/rs/developing/crdbs/developing-streams-crdb.md:
```diff
@@ -1,278 +1,265 @@
+1. XDEL, XCLAIM, and other commands may affect more than one entry when duplicate IDs are present
+1. Entries with duplicate IDs may be removed if a database is exported or renamed
+
+To change XADD's ID mode, use the `rlutil` command-line utility:
```
Will repleace with rladmin commands.

banker commented:
@Paz-Yanover @AlonMagrafta Can you review the latest changes and approve?

Paz-Yanover commented on content/rs/developing/crdbs/developing-streams-crdb.md:
````diff
@@ -1,278 +1,320 @@
-            2) "v"
+Set _strict_ mode:
+```sh
+rladmin> tune db crdb crdt_xadd_id_uniqueness_mode strict
````
I suggest writing it in the following format:
rladmin> tune db <crdb name> crdt_xadd_id_uniqueness_mode strict

AlonMagrafta replied:
I find both confusing. I suggest:
`rladmin> tune myCRDB crdt_xadd_id_uniqueness_mode strict`

Paz-Yanover commented on content/rs/developing/crdbs/developing-streams-crdb.md:
```diff
@@ -1,278 +1,320 @@
+| _t1_   | `XADD x 110 f1 v1`                                   | `XADD x 115 f1 v1`                                   |
+| _t2_   | `XADD x 120 f1 v1`                                   |                                                    |
+| _t3_   | `XADD x 130 f1 v1`                                   |                                                    |
+| _t4_   | `XREAD COUNT 2 STREAMS x 0` <br/>**→ [110-2700, 120-2700]**       |                                                    |
```
Any reason for using so large seq. IDs? I find it confusing

AlonMagrafta replied:
+ my comment on the SEQ part (need to be %32)

Paz-Yanover commented on content/rs/developing/crdbs/developing-streams-crdb.md:
```diff
@@ -1,278 +1,320 @@
 
-In consumer groups, every XREADGROUP and every XACK can change the state of a group or consumer,
-but this cross-location replication is slow and bandwidth is limited.
+Calls to XREADGROUP and XACK change the state of a group or consumer. However, it's not efficient to replicate every change to a consumer or group.
```
The term is "consumer group", I would avoid calling it just a group

Paz-Yanover commented on content/rs/developing/crdbs/developing-streams-crdb.md:
```diff
@@ -1,278 +1,320 @@
-XREADGOUP does not "miss" entries, whereas XREAD can "miss" entries when used in a SCAN-like pattern with concurrent XADD operations.
-In traffic redirection, XREADGOUP can return entries that were read but not acknowledged.
+Unlike XREAD, XREADGOUP will never skip stream entries.
+In traffic redirection, XREADGROUP may return entries that have been read but not acknowledged.
```
It may also return entries that have been acked (consecutive ack issue)

AlonMagrafta commented on content/rs/developing/crdbs/developing-streams-crdb.md:
```diff
@@ -1,278 +1,320 @@
-| t6   | XRANGE x - + --> [100-2] | XRANGE x - + --> [100-2] |
+<tr>
+<td><em>t3</em></td>
+<td><code>XRANGE messages - +</code> <br/><strong>→ [1589929244828-2100]</strong></code></td>
```
One more thing - 
The SEQ part (which is 2100, 3700 here) should be the `SEQ == instance_number%32`. So I would expect 1,2.
Moveover - their %32 for both is 20 which can't be in strict mode (this would mean they are both the same instance)

banker replied:
@AlonMagrafta Not sure what you mean by: "their %32 for both is 20" where do you see this?

AlonMagrafta replied:
What I meant was
2100%32 = 20 as well as 3700%32 = 20
Which indicates both are the same region. But I see you updated it to be 1 and 2 which is the right

AlonMagrafta commented on content/rs/developing/crdbs/developing-streams-crdb.md:
````diff
@@ -1,278 +1,320 @@
-Using XREAD is not recommended as an iterator in Active-Active databases because entries can be skipped:
+Set _semi-strict_ mode:
+```sh
+rladmin> tune db crdb crdt_xadd_id_uniqueness_mode semi-strict
````
here too: change db name from "crdb" to "myCRDB"

AlonMagrafta commented on content/rs/developing/crdbs/developing-streams-crdb.md:
````diff
@@ -1,278 +1,320 @@
+
+Set _liberal_ mode:
+```sh
+rladmin> tune db crdb crdt_xadd_id_uniqueness_mode liberal
````
here too: change db name from "crdb" to "myCRDB"

AlonMagrafta commented on content/rs/developing/crdbs/developing-streams-crdb.md:
```diff
@@ -1,278 +1,320 @@
-As shown in the previous example, Location 2 is not aware of any group-related activity and it seems that redirecting the XREADGROUP traffic from Location 1 to Location 2 results in reading the same entries that were already read.
+Using XREADGROUP across instances can result in instances reading the same entries.
+This is due to the fact that Active-Active Streams is designed for at-least-once reads or a single consumer.
+As shown in the previous example, Region 2 is not aware of any group-related activity, so redirecting the XREADGROUP traffic from Region 1 to Region 2 results in reading entries that have already been read.
```
as Paz said - "consumer group" instead of "group"

banker commented on content/rs/developing/crdbs/developing-streams-crdb.md:
```diff
@@ -1,278 +1,320 @@
 ---
-Title: Developing with Streams in a Active-Active databases
+Title: Streams with Active-Active databases
```
I find the "Developing with..." unnecessarily verbose. I will change the titles of the other articles.

banker commented:
@Paz-Yanover @AlonMagrafta I believe I've addressed most of the issues here. As for the long millisecond value in some IDs, this is indeed how server-generated IDs appear. Any, please take a quick look. And @AlonMagrafta, please see my clarification question.

AlonMagrafta commented:
I pushed one more commit with small IDs fixes

banker commented:
@AlonMagrafta The reason used "region" over "instance" is that @Paz-Yanover suggested it. Personally, I think that region is easier to conceptualize. What's more, the word "region" more accurately reflects that actual usage of active-active. Please advise so we can close and merge this issue.
